source/drivers/HIDJoystick.cpp

```diff
@@ -0,0 +1,204 @@
+DEALINGS IN THE SOFTWARE.
+*/
+
+#include "HIDJoystick.h"
+#include "HID.h"
+
+#if CONFIG_ENABLED(DEVICE_USB)
+
+using namespace codal;
+
+//this descriptor must be stored in RAM
+static char HIDJoystickDescriptor[] = {
+	 0x05, 0x01, // USAGE_PAGE (Generic Desktop)
+	 0x09, 0x05, // USAGE (Game Pad)
+	 0xa1, 0x01, // COLLECTION (Application)
+	 0x05, 0x02, // USAGE_PAGE (Simulation Controls)
+	 0x09, 0xbb, // USAGE (Throttle)
+	 0x15, 0x00, // LOGICAL_MINIMUM (0)
+	 0x25, 0x1f, // LOGICAL_MAXIMUM (31)
+	 0x75, 0x08, // REPORT_SIZE (8)
+	 0x95, 0x01, // REPORT_COUNT (1)
+	 0x81, 0x02, // INPUT (Data,Var,Abs)
+	 0x05, 0x02, // USAGE_PAGE (Simulation Controls)
+	 0x09, 0xbb, // USAGE (Throttle)
+	 0x15, 0x00, // LOGICAL_MINIMUM (0)
+	 0x25, 0x1f, // LOGICAL_MAXIMUM (31)
+	 0x75, 0x08, // REPORT_SIZE (8)
+	 0x95, 0x01, // REPORT_COUNT (1)
+	 0x81, 0x02, // INPUT (Data,Var,Abs)
+	 0x05, 0x01, // USAGE_PAGE (Generic Desktop)
+	 0xa1, 0x00, // COLLECTION (Physical)
+	 0x09, 0x30, // USAGE (X)
+	 0x09, 0x31, // USAGE (Y)
+	 0x09, 0x32, // USAGE (Z)
+	 0x09, 0x33, // USAGE (Rx)
+	 0x15, 0x81, // LOGICAL_MINIMUM (-127)
+	 0x25, 0x7f, // LOGICAL_MAXIMUM (127)
+	 0x75, 0x08, // REPORT_SIZE (8)
+	 0x95, 0x04, // REPORT_COUNT (4)
+	 0x81, 0x02, // INPUT (Data,Var,Abs)
+	 0x05, 0x09, // USAGE_PAGE (Button)
+	 0x19, 0x01, // USAGE_MINIMUM (Button 1)
+	 0x29, 0x10, // USAGE_MAXIMUM (Button 16)
+	 0x15, 0x00, // LOGICAL_MINIMUM (0)
+	 0x25, 0x01, // LOGICAL_MAXIMUM (1)
+	 0x75, 0x01, // REPORT_SIZE (1)
+	 0x95, 0x10, // REPORT_COUNT (16)
+	 0x81, 0x02, // INPUT (Data,Var,Abs)
+	 0xc0, // END_COLLECTION
+	 0xc0 // END_COLLECTION
+};
+
+static const HIDReportDescriptor reportDesc = {
+	9,
+	0x21,                  // HID
+	0x101,                 // hidbcd 1.01
+	0x00,                  // country code
+	0x01,                  // num desc
+	0x22,                  // report desc type
+	sizeof(HIDJoystickDescriptor),
+};
+
+static const InterfaceInfo ifaceInfo = {
+	&reportDesc,
+	sizeof(reportDesc),
+	1,
+	{
+		1,    // numEndpoints
+		0x03, /// class code - HID
+		0x01, // subclass (boot interface)
+		0x02, // protocol (joystick)
+		0x00, //
+		0x01, //
+	},
+	{USB_EP_TYPE_INTERRUPT, 1},
+	{USB_EP_TYPE_INTERRUPT, 1},
+};
+
+static HIDJoystickState joystickState = {
+	0, 0, 0, 0, 0, 0, 0,
+};
+
+USBHIDJoystick::USBHIDJoystick() : USBHID()
+{
+
+}
+
+int USBHIDJoystick::stdRequest(UsbEndpointIn &ctrl, USBSetup &setup)
+{
+	if (setup.bRequest == GET_DESCRIPTOR)
+	{
+		if (setup.wValueH == 0x21)
+		{
+			InterfaceDescriptor tmp;
+			fillInterfaceInfo(&tmp);
+			return ctrl.write(&tmp, sizeof(tmp));
+		}
+		else if (setup.wValueH == 0x22)
+		{
+			return ctrl.write(HIDJoystickDescriptor, sizeof(HIDJoystickDescriptor));
+		}
+	}
+	return DEVICE_NOT_SUPPORTED;
+}
+
+const InterfaceInfo *USBHIDJoystick::getInterfaceInfo()
+{
+	return &ifaceInfo;
+}
+
+int USBHIDJoystick::buttonDown(uint8_t b)
+{
+	uint16_t btn = (1UL << b);
```

Comment:
Is there a common subset of buttons for joystick controllers? 65535 buttons seems quite a lot!

Reply:
Also, a function like `click(uint8_t button)` might be nice.

Reply:
there are 16 buttons in this implementation. A 16 bit integer is used to hold this information with 1 bit per button. I am not currently aware of any standardized layouts, but this layout seems to make sense when imagining a current xbox controller (2x bumpers, start, select. abxy, start, select, click stick 1, click stick 2 = 12 buttons, next size up integer is 16 bits).

I can add a higher level click function as well with builtin delay.

Reply:
LOL, silly me. 

Ok, seems like in the future we should add a few standard subclasses for an Xbox controller etc.